File: Top200/213-Rob.java
package Top200;

class Rob {
    public int rob(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }

        if (nums.length == 1) {
            return nums[0];
        }

        if (nums.length == 2) {
            return Math.max(nums[0], nums[1]);
        }

        // dp 数组
        int[] dp1 = new int[nums.length - 1];
        int[] dp2 = new int[nums.length - 1];
        
        // dp1 初始化
        dp1[0] = nums[0];
        dp1[1] = Math.max(nums[0], nums[1]);

        dp2[0] = nums[1];
        dp2[1] = Math.max(nums[1], nums[2]);

        //
        for (int i = 2; i < nums.length - 1; i++) {
            dp1[i] = Math.max(dp1[i-1], dp1[i-2] + nums[i]);
        }

        for (int i = 3; i < nums.length; i++) {
            dp2[i-1] = Math.max(dp2[i-2], dp2[i-3] + nums[i]);
        }

        return Math.max(dp1[nums.length - 2], dp2[nums.length - 2]);
    }
}

File: Top200/213-Rob.go
package main

import "math"

func rob(nums []int) int {
	l := len(nums)
	if l == 0 {
		return 0
	}

	if l == 1 {
		return nums[0]
	}

	if l == 2 {
		return int(math.Max(float64(nums[0]), float64(nums[1])))
	}

	dp1 := make([]int, l)
	dp2 := make([]int, l)

	dp1[0] = nums[0]
	dp1[1] = int(math.Max(float64(nums[0]), float64(nums[1])))

	dp2[0] = 0
	dp2[1] = nums[1]
	dp2[2] = int(math.Max(float64(nums[1]), float64(nums[2])))

	for i := 2; i < l-1; i++ {
		dp1[i] = int(math.Max(float64(dp1[i-1]), float64(dp1[i-2]+nums[i])))
		dp2[i+1] = int(math.Max(float64(dp2[i]), float64(dp2[i-1]+nums[i+1])))
	}

	return int(math.Max(float64(dp1[l-2]), float64(dp2[l-1])))
}
